### ISSUE TITLE
bug: Wrong CWD When Using Dev Containers on Mac

### ISSUE THREAD

**Issue by mharradon** [vscode-neovim/vscode-neovim#2258]
### Check the following:

- [X] I have tried setting `vscode-neovim.neovimClean` in VSCode settings and **restarting VSCode.**
- [X] I have read the vscode-neovim docs.
- [X] I have searched existing (closed!) issues.

### Neovim version (nvim -v)

v0.10.1

### Operating system/version

Mac 14.4.1

### Problem (including [log messages](https://github.com/vscode-neovim/vscode-neovim?tab=readme-ov-file#troubleshooting))

On a fresh setup with a Dev Container Neovim sets the CWD to `/`. This prevents e.g. tab path completion working when opening new files.

### Steps To Reproduce

1) Open a project with Neovim and Dev Container Extensions
2) `:` followed by `lua print(vim.fn.getcwd())`
3) Output is `/`

### Expected Behavior

I expect the reported working directory to be the root of the project I am working in.

**Comment by xiyaowong** [comment]
Is vscode-neovim installed in a remote environment? (I mean the vscode-neovim extension, not neovim itself.)

**Comment by mharradon** [comment]
No, it's installed locally.

**Comment by xiyaowong** [comment]
https://github.com/vscode-neovim/vscode-neovim?tab=readme-ov-file#remote-development

**Comment by mharradon** [comment]
I'm trying to understand your response. I want to use vscode-neovim as a
local UI extension, so it is installed locally, and should "work out of the
box", as the docs you linked states. Is your reply intended to imply this
is broken and I should use remote Neovim?

**Comment by xiyaowong** [comment]
How could the CWD be a remote location if the extension runs locally?

**Comment by mharradon** [comment]
The CWD is just the diagnosable issue. All I want is tabbed path completion
to work properly. Presumably using the local CWD over a random path would
be part of a solutiomn.

At a minimium if it's not possible for this UI extension to support the
standard vim/neovim methods for opening files when used as a front-end
extension I think that would be a good thing to document. Presumably
opening files is a commonly-used feature.

**Comment by justinmk** [comment]
We definitely should try to set a useful CWD. Is there a reason we can't?

**Comment by justinmk** [comment]
> All I want is tabbed path completion to work properly

Are you hitting `<tab>` in Nvim? Describe the steps. 

If Nvim is running locally in a remote vscode session, then setting CWD won't help with tab-completion, because Nvim doesn't know about the vscode VFS. But depending on the steps you describe, there may be other things that can work.

**Comment by mharradon** [comment]
1) Open a project with a Dev Container. Presume it has a.txt and b.txt in its root.
2) Open the a.txt file
3) Type :tabe b<TAB>

Expectation: Presented with auto complete menu including b.txt

Result: Presented with files starting with `b` in my machine root directory. 

This does work as expected when not using devcontainers. 

It's a workflow of particular interest to me because I'm only using VSCode in the first place to experiment with using Dev Containers while maintaining NeoVim ergonomics.

Maybe this is just not possible in this configuration, but hypothetically it seems possible to set CWD to the local project and then intercept resulting fopen etc. commands and redirect them to whatever remote equivalent VSCode expects. That's assuming there's not already some 'virtual directory structure' VSCode creates locally that could be used directly.

**Comment by justinmk** [comment]
> but hypothetically it seems possible to set CWD to the local project and then intercept resulting fopen etc. commands and redirect them to whatever remote equivalent VSCode expects

That would be the VFS concept I mentioned. Nvim currently doesn't have a VFS interface (indirectly discussed in https://github.com/neovim/neovim/issues/21635 , just vaporware currently), but that would be something that Nvim should provide.

Meanwhile, if `nvim` is [running on the remote](https://github.com/vscode-neovim/vscode-neovim?tab=readme-ov-file#remote-development) then yes, we should ensure that vscode-neovim sets a useful CWD there. If it's running locally, then no, it can't work without a VFS layer.

**Comment by mharradon** [comment]
Thank you, that makes sense. I will try running as a remote extension as a work-around.

I think I'm suggesting something slightly different than a VFS interface for Nvim. Rather, let Nvim use the local project CWD. Path completion should work reasonably in that case. Then when translating open-file calls to VSCode APIs change the paths to refer to remote URIs corresponding to the open remote workspace. The API calls vscode-neovim uses now take URIs, and it appears those URIs can reference the remote filesystem (e.g. https://stackoverflow.com/questions/77801691/how-to-open-vscode-with-vscode-vscode-remote-with-devcontainers).

**Comment by justinmk** [comment]
> I think I'm suggesting something slightly different than a VFS interface for Nvim. Rather, let Nvim use the local project CWD. Path completion should work reasonably in that case.

But that would only work in the rather special case where the local `nvim` CWD happens to point to a project that "looks like" whatever workspace is on the remote?

**Comment by mharradon** [comment]
I also think an approach like I describe would resolve a variety of similar issues WSL users have.

If PRs in this direction would be welcome please let me know.

**Comment by justinmk** [comment]
> If PRs in this direction would be welcome please let me know.

Sure, it would help! Start with a quick hack/prototype for discussion purposes.

**Comment by theol0403** [comment]
@mharradon I think your suggested approach is impossible without a VFS. If your code "lives" in a remote container, but nvim is running on your local machine... then nvim has NO access to the real files on the remote, and instead only sees buffer contents (opened from vscode).

Yes, we can translate open/save URIs so that `:e remote_file.txt` works, but we cannot do anything about tab completion, because this is done internally within nvim (ie there is no vim.fs.ls() or autocommand we can stub).

TLDR the only way nvim can have any understanding of the remote file tree for the purposes of tab completion, netrw, etc, is by integrating nvim<->vscode through a VFS. 

If I'm not missing anything I suggest we close this issue until nvim comes up with a new API.

**Comment by justinmk** [comment]
This issue has good info though so we can mark it as blocked and keep it open?
